sdk/nodejs/provisioning/cli-downloader.ts

```diff
@@ -0,0 +1,282 @@
+import AdmZip from "adm-zip"
+import * as crypto from "crypto"
+import envPaths from "env-paths"
+import * as fs from "fs"
+import fetch from "node-fetch"
+import * as os from "os"
+import * as path from "path"
+import * as tar from "tar"
+
+import { InitEngineSessionBinaryError } from "../common/errors/index.js"
+
+const CLI_HOST = "dl.dagger.io"
+const DAGGER_CLI_BIN_PREFIX = "dagger"
+const CACHE_DIR = path.join(
+  `${process.env.XDG_CACHE_HOME?.trim() || envPaths("", { suffix: "" }).cache}`,
+  "dagger"
+)
+
+interface CliDownloaderOptions {
+  cliVersion: string
+  archive?: {
+    checksumUrl?: string
+    url?: string
+    name?(architecture: string): string
+    extract?(archivePath: string, destinationPath: string): void
+    path?(destinationFolder: string): string
+  }
+  executableFilename?(name: string): string
+}
+
+export class CliDownloader {
+  private readonly cliVersion: string
+  private readonly archive?: CliDownloaderOptions["archive"]
+  private readonly executableFilename: (name: string) => string
+
+  constructor(options: CliDownloaderOptions) {
+    if (!options.cliVersion) {
+      throw new Error("cliVersion is not set")
+    }
+
+    this.cliVersion = options.cliVersion
+    this.archive = options.archive
+    this.executableFilename =
+      options.executableFilename ?? ((name: string) => name)
+  }
+
+  static async Download(options: CliDownloaderOptions): Promise<string> {
+    const cliDownloader =
+      os.platform() === "win32"
+        ? new WindowsCliDownloader(options)
+        : new CliDownloader(options)
+
+    return await cliDownloader.Download()
+  }
+
+  async Download(): Promise<string> {
```

Comment:
@TomChv Just referencing your comment from the previous PR, I tend to like pascalCase method names in javascript files, but I was following, what appeared to be, a convention of public methods being PascalCase vs camelCase for private methods. Maybe that was an incorrect assumption on my part?

Reply:
AFAIK, Class methods are using lowerCamelCase by convention.
- https://google.github.io/styleguide/jsguide.html#naming-method-names
- https://github.com/airbnb/javascript#naming--camelCase

We should stay consistent in capitalization with the rest of our codebase. 
Formatting is still pretty subjective though ;)
- https://github.com/ryanmcdermott/clean-code-javascript#formatting

Reply:
I'm totally down for that.

For future reference, the other classes in provisioning folder will need updated for consistency.  For instance, https://github.com/dagger/dagger/blob/main/sdk/nodejs/provisioning/bin.ts

Reply:
100% agreed